### PR TITLE
✨ Implement retry strategy

### PR DESCRIPTION
**Pull request by bisgaard-itis**
- Add a proper retry strategy to retry requests if the server responds with status in {429, 503, 504}.
- Note that the core changes have actually been done to the auto generation tool we are using: https://github.com/ITISFoundation/openapi-generator/commit/cbddf820750bd3fba7c8ee22ee6c509b6fed5007 https://github.com/ITISFoundation/openapi-generator/commit/1f5e6ddf02364d3189a09fe090b2dfcd3f0df5c1. This simply propagates these changes to the hand-written part of the code.
- I ran these changes on the e2e tests which passed...

## Related issue/s
- resolves https://github.com/ITISFoundation/osparc-simcore-clients/issues/134

### CODE REVIEW

File: clients/python/client/osparc/_http_client.py [hunk 1]
```diff
@@ -41,3 +44,44 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> None:
...
+    ) -> httpx.Response:
+        @tenacity.retry(
+            reraise=True,
+            wait=tenacity.wait_exponential(
+                multiplier=self.configuration.retries.backoff_factor
+            ),
+            stop=tenacity.stop_after_attempt(
+                self.configuration.retries.total
+                if isinstance(self.configuration.retries.total, int)
+                else 4
```
Comment (pcrespov) [inline clients/python/client/osparc/_http_client.py:59]: No magic numbers please. Create a constant for this. :-)

File: clients/python/client/osparc/_http_client.py [hunk 2]
```diff
@@ -41,3 +44,44 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> None:
...
+            ),
+            retry=tenacity.retry_if_exception_type(httpx.HTTPStatusError),
+        )
+        async def _():
+            response: httpx.Response = await method(*args, **kwargs)
+            try:
+                response.raise_for_status()
```
Comment (pcrespov) [inline clients/python/client/osparc/_http_client.py:66]: i think here you can just check with an `if `. Actually this is what they do in the implementaiton of that function. SEE
https://github.com/ITISFoundation/osparc-issues/issues/1136#issuecomment-1943836871

File: clients/python/client/osparc/_http_client.py [hunk 3]
```diff
@@ -41,3 +44,44 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> None:
                     raise err from exc_value
             await self._client.aclose()
             raise exc_value
+
+    async def _request(
+        self, method: Callable[[Any], Awaitable[httpx.Response]], *args, **kwargs
+    ) -> httpx.Response:
+        @tenacity.retry(
```
Comment (pcrespov) [inline clients/python/client/osparc/_http_client.py:51]: What about the case in which you have a 503 with retry-after header in place? perhaps you should add a tenacity hook for that case?

**Review by pcrespov — APPROVED** [review]
Looks good! thx
I left some thoughts a suggestions.

File: clients/python/client/osparc/_http_client.py [hunk 4]
```diff
@@ -41,3 +44,44 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> None:
...
+            reraise=True,
+            wait=tenacity.wait_exponential(
+                multiplier=self.configuration.retries.backoff_factor
```
[the remark this answers is not in the document]
Comment (bisgaard-itis) [reply clients/python/client/osparc/_http_client.py:54]: The `retries.backoff_factor` was added in this commit: https://github.com/ITISFoundation/openapi-generator/commit/cbddf820750bd3fba7c8ee22ee6c509b6fed5007
That commit is the main contribution which should ensure the auto generated client will retry. This PR only handles the thin hand-written wrapper. (where we use a `httpx` client).

File: clients/python/client/osparc/_http_client.py (hunk 1 above)
Comment (bisgaard-itis) [reply clients/python/client/osparc/_http_client.py:59]: Right, the total should be an `int` so I think I will just add an assert and avoid adding an extra constant here all together.

File: clients/python/client/osparc/_http_client.py (hunk 2 above)
Comment (bisgaard-itis) [reply clients/python/client/osparc/_http_client.py:66]: Good point. I will correct.

Comment (bisgaard-itis) [reply clients/python/client/osparc/_http_client.py:67]: Good point, I fixed this

File: clients/python/client/osparc/_http_client.py (hunk 3 above)
Comment (bisgaard-itis) [reply clients/python/client/osparc/_http_client.py:58]: Yep, I added this case along with some tests as well

**Pull request merged**